### PR TITLE
Initial code for Hackathon

### PR DESCRIPTION
**Pull request by tomergafner**

### CODE REVIEW

File: hackathon_example.py [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+    (
+        model_input_image,
+        predictions,
+        expected_label,
+        load_pre_trained_weights_op,
+        train_op,
+        metrics_to_values,
+        metrics_to_updates,
+    ) = initialize_pre_trained_inception_v1()
+
```
Comment (tomergafner) [inline hackathon_example.py]: local_init_op = tf.local_variables_initializer()

File: hackathon_example.py [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+        training_accuracy = []
+        training_loss = []
+        validation_accuracy = []
+        validation_loss = []
+        best_validation_accuracy = None
+
+        # Running training loop.
+        for _ in range(number_of_epochs):
+            session.run(tf.local_variables_initializer())
```
Comment (tomergafner) [inline hackathon_example.py]: `session.run(local_init_op)`

File: hackathon_example.py [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+    validation_iterator = create_dataset_iterator(
+        image_path_placeholder,
+        label_placeholder,
+        batch_size,
+    )
+    next_validation_batch = validation_iterator.get_next()
+
+    # Define inception v1 and return ops to load pre-trained model (trained on
+    # ImageNet).
+    (
```
Comment (tomergafner) [inline hackathon_example.py]: if you call out to a new func, consider returning a `tf.train.Scaffold` which is a collection of stuff for training.

File: hackathon_example.py [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+@click.option(
+    '--dest_dir',
+    required=True,
+    default=os.getcwd(),
+    type=click.Path(exists=True, dir_okay=True)
+)
+@cli.command()
+def download_dataset(dest_dir):
+    """
+    Download and extract the mini dataset.
```
Comment (nathansilberman) [inline hackathon_example.py]: Inline this:

"""Download and extract...

File: hackathon_example.py [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+    dataset_image_paths, dataset_labels = load_data_from_csv(input_file)
+
+    with tf.Graph().as_default() as graph:
+
+        # Define the data iterator.
+        image_paths = tf.placeholder(tf.string, [None])
+        label_data = tf.placeholder(tf.int32, [None])
+
+        test_iterator = create_dataset_iterator(
+            image_paths,
```
Comment (nathansilberman) [inline hackathon_example.py]: one line:

```
test_iterator = create_dataset_iterator(
    image_paths, label_data, batch_size,
)
```

File: hackathon_example.py [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+        [None, TARGET_IMAGE_WIDTH, TARGET_IMAGE_HEIGHT, 3],
+        name="images"
+    )
+    labels = tf.placeholder(tf.int64, [None], name="labels")
+
+    # Define inception v1 and return ops to load pre-trained model (trained on
+    # ImageNet).
+
+    restore_op, feed_dict, train_op, metrics_to_values, metrics_to_updates \
+        = initialize_pre_trained_inception_v1(images, labels)
```
Comment (nathansilberman) [inline hackathon_example.py]: just call this function

```
= create_mode(images, labels)
```

File: hackathon_example.py [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+    """
+
+    train_image_paths, train_labels, valid_image_paths, valid_labels = \
+        load_data_from_csv(input_file, split=True)
+
+    # Define the data iterators.
+    image_paths = tf.placeholder(tf.string, [None])
+    label_data = tf.placeholder(tf.int32, [None])
+
+    training_iterator = create_dataset_iterator(
```
Comment (nathansilberman) [inline hackathon_example.py]: one line: `image_paths, label_data, batch_size`, same below.

File: hackathon_example.py [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+        optimizer = tf.train.AdamOptimizer(1e-4)
+
+        # Create a train op.
+        train_op = tf.contrib.training.create_train_op(
+            loss,
+            optimizer,
+            variables_to_train=variables_to_train,
+        )
+
+        predictions = tf.identity(
```
Comment (nathansilberman) [inline hackathon_example.py]: You shouldn't need to use identity here, just call:

```
= tf.argmax(end_points['Predictions'], 1, name='predictions')
```

File: hackathon_example.py [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+            tf.argmax(end_points['Predictions'], 1),
+            name="predictions"
+        )
+        metrics_to_values, metrics_to_updates = \
+            slim.metrics.aggregate_metric_map({
+                'accuracy': tf.metrics.accuracy(
```
Comment (nathansilberman) [inline hackathon_example.py]:
```
'accuracy': tf.metrics.accuracy(predictions, labels),
'mean_loss': ...
```

File: hackathon_example.py [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+            # Save model if accuracy improved.
+            if (
+                    (not best_validation_accuracy) or
+                    best_validation_accuracy < accuracy
+            ):
+                best_validation_accuracy = accuracy
+                saver.save(session, os.path.join(export_dir, 'buttefly-model'))
+
+
+def initialize_pre_trained_inception_v1(images, labels):
```
Comment (nathansilberman) [inline hackathon_example.py]: add a docstring, describe what you're returning.

File: hackathon_example.py [hunk 11]
```diff
@@ -0,0 +1,493 @@
...
+    python hackathon_example.py download_dataset
+    """
+    mini_dataset_url = \
+        'https://github.com/ButterflyNetwork/YaleHackathon2018' \
+        '/releases/download/v.0.0.1/butterfly_mini_dataset.tar.gz'
+
+    filename = os.path.basename(mini_dataset_url)
+    filepath = os.path.join(dest_dir, filename)
+    if not os.path.exists(filepath):
+        def _progress(count, block_size, total_size):
```
Comment (imalkin) [inline hackathon_example.py]: I prefer tqdm, but I think it's a style question.

File: hackathon_example.py [hunk 12]
```diff
@@ -0,0 +1,494 @@
...
+            validation_set['label'].tolist(),
+        )
+    else:
+        return (
+            df['image_file_path'].tolist(),
+            df['label'].tolist(),
+        )
+
+
+def load_image(filepath, label):
```
Comment (nathansilberman) [inline hackathon_example.py]: the label is just passed through - why have it be an argument at all?

```
def load_image(path):
   ...
  return image
```

File: hackathon_example.py [hunk 13]
```diff
@@ -0,0 +1,493 @@
...
+                            image_paths: dataset_image_paths,
+                            label_data: dataset_labels,
+                        })
+
+            while True:
+                try:
+                    # Read the next batch.
+                    batch_images, batch_labels = session.run(next_test_batch)
+
+                    # Train the model.
```
Comment (imalkin) [inline hackathon_example.py]: Evaluating, not training.

File: hackathon_example.py [hunk 14]
```diff
@@ -0,0 +1,493 @@
...
+@click.option('--number_of_epochs', required=False, type=click.INT, default=10)
+@click.option(
+    '--export_dir',
+    required=True,
+    type=click.Path(exists=False, dir_okay=True)
+)
+def train(input_file, batch_size, number_of_epochs, export_dir):
+    """
+    :param input_file: the csv file containing the training set.
+    :param batch_size: the batch size used for training
```
Comment (imalkin) [inline hackathon_example.py]: nit: period at end of all param descriptions.

File: hackathon_example.py (hunk 12 above)
Comment (tomergafner) [reply hackathon_example.py]: This what the function used for mapping requires.

File: hackathon_example.py (hunk 11 above)
Comment (tomergafner) [reply hackathon_example.py]: will change Thanks !
